mempool: Log added for dumping mempool transactions to disk

kevkevinpal opened the pull request:
Sometimes when shutting off bitcoind it can take a while to dump the mempool transaction onto the disk so
this change adds additional logging to the `DumpMempool` method in `kernel/mempool_persist.cpp`

Motivated by https://github.com/bitcoin/bitcoin/pull/29227 this change
 - adds a single new line for the amount of transactions being dumped and the amount of memory being dumped to file

This is in response to https://github.com/bitcoin/bitcoin/pull/29227#issuecomment-1893375082

The logs will now look like this
```
2024-02-09T23:41:52Z DumpAnchors: Flush 2 outbound block-relay-only peer addresses to anchors.dat completed (0.02s)
2024-02-09T23:41:52Z scheduler thread exit
2024-02-09T23:41:52Z Writing 29 mempool transactions to file...
2024-02-09T23:41:52Z Writing 0 unbroadcast transactions to file.
2024-02-09T23:41:52Z Dumped mempool: 0.000s to copy, 0.022s to dump, 0.015 MB dumped to file
2024-02-09T23:41:52Z Flushed fee estimates to fee_estimates.dat.
2024-02-09T23:41:53Z Shutdown: done
```

kevkevinpal committed:
mempool: Log added for dumping mempool transactions to disk

kevkevinpal commented:
> Why is this needed? A progress log for something that takes less than a second does not sound useful

This is in response to this comment https://github.com/bitcoin/bitcoin/pull/29227#pullrequestreview-1817064507 
the logs I posted in the description are from testnet I can run this on mainnet to get the time it takes on shutdown there

epiccurious commented:
> logging the detailed progress in % is a bit overkill

+1 to @theStack's comment. Logging the start and end would be sufficient, no need for the 10% increments.

kevkevinpal commented:
force pushed [33153b0](https://github.com/bitcoin/bitcoin/pull/29402/commits/33153b0236346809c75f6f52a3000a4e18e3b385)

This removes the 10% increments and now logs a single line with the amount of mempool transactions being written to the disk

maflcko reviewed (approved):
lgtm

kevkevinpal commented on src/kernel/mempool_persist.cpp:
```diff
@@ -184,7 +184,9 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
-        file << (uint64_t)vinfo.size();
+        uint64_t mempool_transactions_to_write = (uint64_t)vinfo.size();
+        file << mempool_transactions_to_write;
+        LogInfo("Writing %u mempool transactions to disk...\n", mempool_transactions_to_write);
```
updated in [c23f5be](https://github.com/bitcoin/bitcoin/pull/29402/commits/c23f5beba2fbcea995b5dc7cab7c88fe0a2ecd22)
I just used `sizeof()` on `vinfo` to get this amount in bytes then multiplied by 1000000 to get MB

kevkevinpal commented on src/kernel/mempool_persist.cpp:
```diff
@@ -184,7 +184,9 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
         file.SetXor(xor_key);
 
-        file << (uint64_t)vinfo.size();
+        uint64_t mempool_transactions_to_write = (uint64_t)vinfo.size();
```
fix in [c23f5be](https://github.com/bitcoin/bitcoin/pull/29402/commits/c23f5beba2fbcea995b5dc7cab7c88fe0a2ecd22)

kevkevinpal commented on src/kernel/mempool_persist.cpp:
```diff
@@ -184,7 +184,9 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
-        file << (uint64_t)vinfo.size();
+        uint64_t mempool_transactions_to_write = (uint64_t)vinfo.size();
+        file << mempool_transactions_to_write;
+        LogInfo("Writing %u mempool transactions to disk...\n", mempool_transactions_to_write);
```
should I update `disk` to `file` in this PR? I was trying to write the logs similar to the other ones in this file

kevkevinpal replied:
added this commit [9f57a2e](https://github.com/bitcoin/bitcoin/pull/29402/commits/9f57a2ec49bf2c298acccbf8d889e692bb6894f7) to change the wording from `disk` to `file`, disk is used in the file once still but I think it makes sense as it is "opened mempool file from disk"

epiccurious commented:
utACK 6312513b5074f160971ffec0b093b58b9d3cdaae. Planning to test as well.

kevkevinpal committed:
log: renamed disk to file so wording was more accurate

epiccurious commented:
I'm seeing an issue during testing. The debug log incorrectly reports 0 MB. Are others able to reproduce my result?

```
user1@comp1:~/Documents/GitHub/btc29402$ src/bitcoin-cli --rpcwait getmempoolinfo | jq -c '{ transactions: (.size), memory_usage_in_mb: (.usage/1000000) }' | jq; src/bitcoin-cli --rpcwait stop; sleep 10; grep -a Writing ~/.bitcoin/debug.log | tail -2
{
  "transactions": 2529,
  "memory_usage_in_mb": 6.061648
}
Bitcoin Core stopping
2024-02-09T14:11:49Z Writing 2529 mempool transactions to file 0 MB...
2024-02-09T14:11:49Z Writing 0 unbroadcast transactions to file.
```

epiccurious commented on src/kernel/mempool_persist.cpp:
```diff
@@ -184,7 +184,9 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
-        file << (uint64_t)vinfo.size();
+        uint64_t mempool_transactions_to_write(vinfo.size());
+        file << mempool_transactions_to_write;
+        LogInfo("Writing %u mempool transactions to file %u MB...\n", mempool_transactions_to_write, sizeof(vinfo)/1000000);
```
very minor nit - To make the log more readable, can you please add a comma after the 'file'?

So it displays as:
```
Writing 2529 mempool transactions to file, xx MB...
```

kevkevinpal commented:
> I'm seeing an issue during testing. The debug log incorrectly reports 0 MB. Are others able to reproduce my result?
> 
> ```
> user1@comp1:~/Documents/GitHub/btc29402$ src/bitcoin-cli --rpcwait getmempoolinfo | jq -c '{ transactions: (.size), memory_usage_in_mb: (.usage/1000000) }' | jq; src/bitcoin-cli --rpcwait stop; sleep 10; grep -a Writing ~/.bitcoin/debug.log | tail -2
> {
>   "transactions": 2529,
>   "memory_usage_in_mb": 6.061648
> }
> Bitcoin Core stopping
> 2024-02-09T14:11:49Z Writing 2529 mempool transactions to file 0 MB...
> 2024-02-09T14:11:49Z Writing 0 unbroadcast transactions to file.
> ```

It looks like `usage` from `getmempoolinfo` is coming from `DynamicMemoryUsage` in `txmempool.cpp`
Where I'm calculating `sizeof(vinfo)` which comes from `infoAll` in `txmempool.cpp`. `infoAll` allocates only `mapTx.size()`

DynamicMemoryUsage computes like this which includes `mapTx.size()`
```
    return memusage::MallocUsage(sizeof(CTxMemPoolEntry) + 15 * sizeof(void*)) * mapTx.size() + memusage::DynamicUsage(mapNextTx) + memusage::DynamicUsage(mapDeltas) + memusage::DynamicUsage(txns_randomized) + cachedInnerUsage;
```

I think it might make more sense to not have a MB amount on `vinfo` and print it out for the file size at the end of the creating the mempool file so users are not confused when they see xx amounts in MB and their file actually a different size

epiccurious commented:
> it might make more sense to not have a MB amount on vinfo and print it out for the file size at the end of the creating the mempool file so users are not confused when they see xx amounts in MB and their file actually a different size

Approach ACK.

kevkevinpal commented:
rebased and also updated code to get the size of the file after saving

logs now look like such and I tested to see that the `mempool.dat` file was the correct size by going to my data dir and doing `ls -alh` and they matched in sizes
```
2024-02-09T23:41:52Z DumpAnchors: Flush 2 outbound block-relay-only peer addresses to anchors.dat completed (0.02s)
2024-02-09T23:41:52Z scheduler thread exit
2024-02-09T23:41:52Z Writing 29 mempool transactions to disk...
2024-02-09T23:41:52Z Writing 0 unbroadcast transactions to file.
2024-02-09T23:41:52Z Dumped mempool: 0.000s to copy, 0.022s to dump, 0.015 MB dumped to file
2024-02-09T23:41:52Z Flushed fee estimates to fee_estimates.dat.
2024-02-09T23:41:53Z Shutdown: done
```

epiccurious commented:
Tested ACK 25482b5aea6fb910b74666414a6330ec53f6a4c9.

This output is clean. Solid approach piggy-backing on the `Dumped mempool: ` line that already exists.
```
2024-02-10T20:36:10Z Writing 5364 mempool transactions to file...
2024-02-10T20:36:10Z Writing 0 unbroadcast transactions to file.
2024-02-10T20:36:10Z Dumped mempool: 0.005s to copy, 0.101s to dump, 7.865 MB dumped to file
```

> they matched in sizes

ACK.
```
user1@comp1:~$ ls -l .bitcoin/mempool.dat 
-rw------- 1 user1 user1 7865003 Feb 10 15:36 .bitcoin/mempool.dat
```

maflcko reviewed (approved):
lgtm

epiccurious commented on src/kernel/mempool_persist.cpp:
```diff
@@ -194,7 +196,7 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
         file << mapDeltas;
 
-        LogPrintf("Writing %d unbroadcast transactions to disk.\n", unbroadcast_txids.size());
+        LogPrintf("Writing %d unbroadcast transactions to file.\n", unbroadcast_txids.size());
```
+1

kevkevinpal replied:
sorry I got confused by the comment and thought you meant that I changed the log line to include `disk` first and then in the second commit changed it to `file`

I updated `LogPrintf` -> `LogInfo` in [27251e8](https://github.com/bitcoin/bitcoin/pull/29402/commits/27251e89b67f83a40cfd70eeded4120a28f82cf9)

kevkevinpal commented on src/kernel/mempool_persist.cpp:
```diff
@@ -205,9 +207,11 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
                   Ticks<SecondsDouble>(mid - start),
-                  Ticks<SecondsDouble>(last - mid));
+                  Ticks<SecondsDouble>(last - mid),
+                  (float)file_size(dump_path)/1000000);
```
I think this is what you are talking about unless I'm getting confused [b9f449b](https://github.com/bitcoin/bitcoin/pull/29402/commits/b9f449b2d0637c09fd8dc3f44078355827af4a92)

the <util/fs.h> includes is already imported but I added `fs::` before `file_size`

kevkevinpal commented on src/kernel/mempool_persist.cpp:
```diff
@@ -205,9 +207,11 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
                   Ticks<SecondsDouble>(mid - start),
-                  Ticks<SecondsDouble>(last - mid));
+                  Ticks<SecondsDouble>(last - mid),
+                  (float)fs::file_size(dump_path)/1000000);
```
instead, I opted not to cast the value and instead print out the number of bytes, let me know if MB is strongly preferred and I can find a solution.

ajtowns reviewed (commented):
utACK dbdb7320252fd68415e8b76bad5a2169bd7da241

kevkevinpal commented on src/kernel/mempool_persist.cpp:
```diff
@@ -44,7 +44,7 @@ bool LoadMempool(CTxMemPool& pool, const fs::path& load_path, Chainstate& active
     AutoFile file{opts.mockable_fopen_function(load_path, "rb")};
     if (file.IsNull()) {
-        LogPrintf("Failed to open mempool file from disk. Continuing anyway.\n");
+        LogInfo("Failed to open mempool file from disk. Continuing anyway.\n");
```
I didn't modify this line because opening the `file from disk` is accurate but I can change this to `Failed to open mempool file. Continuing anyway`

kevkevinpal replied:
updated in 34b6a8dadfb9a279f38fc828d6dff73209928f5d

glozow commented on src/kernel/mempool_persist.cpp:
```diff
@@ -194,7 +196,7 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
 
         file << mapDeltas;
 
-        LogPrintf("Writing %d unbroadcast transactions to disk.\n", unbroadcast_txids.size());
+        LogInfo("Writing %d unbroadcast transactions to file.\n", unbroadcast_txids.size());
```
nit: this seems to be in the wrong commit

kevkevinpal replied:
Updated it is now in 4d5b55735bcf82847d748d24da5dbdbd1de8ef41

glozow reviewed (commented):
utACK 34b6a8dadfb9a279f38fc828d6dff73209928f5d

glozow reviewed (commented):
reACK 4d5b557

tdb3 commented:
crACK will test soon

tdb3 reviewed (commented):
crACK. Will test shortly

tdb3 commented:
Tested from master after merge.  lgtm.

```
2024-03-28T14:37:40Z Writing 21 mempool transactions to file...
2024-03-28T14:37:40Z Writing 0 unbroadcast transactions to file.
2024-03-28T14:37:40Z Dumped mempool: 0.000s to copy, 0.006s to dump, 5795 bytes dumped to file
```

```
2024-03-28T14:37:59Z Loading 21 mempool transactions from file...
...
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 14% (tried 3, 18 remaining)
2024-03-28T14:37:59Z msghand thread start
2024-03-28T14:37:59Z init message: Done loading
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 23% (tried 5, 16 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 33% (tried 7, 14 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 42% (tried 9, 12 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 52% (tried 11, 10 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 61% (tried 13, 8 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 71% (tried 15, 6 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 80% (tried 17, 4 remaining)
2024-03-28T14:37:59Z Progress loading mempool transactions from file: 90% (tried 19, 2 remaining)
2024-03-28T14:37:59Z Imported mempool transactions from file: 21 succeeded, 0 failed, 0 expired, 0 already there, 0 waiting for initial broadcast
```
